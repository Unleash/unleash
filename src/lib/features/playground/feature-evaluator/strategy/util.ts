import * as murmurHash3 from 'murmurhash3js';

function normalizedValue(
    id: string,
    groupId: string,
    normalizer: number,
    seed = 0,
): number {
    const hash = murmurHash3.x86.hash32(`${groupId}:${id}`, seed);
    return (hash % normalizer) + 1;
}

const STRATEGY_SEED = 0;

export function normalizedStrategyValue(id: string, groupId: string): number {
    return normalizedValue(id, groupId, 100, STRATEGY_SEED);
}

const VARIANT_SEED = 86028157;

export function normalizedVariantValue(
    id: string,
    groupId: string,
    normalizer: number,
): number {
    return normalizedValue(id, groupId, normalizer, VARIANT_SEED);
}
