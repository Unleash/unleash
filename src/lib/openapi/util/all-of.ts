import { JSONSchema } from 'json-schema-to-ts';

// this function simplifies simple schemas and return allOf schema if it
// doesn't know how to simplify it. It's a proof of concept but it can be extended
export function mergeAllOf(a: JSONSchema, b: JSONSchema): JSONSchema {
    if (typeof a !== 'boolean' && typeof b !== 'boolean') {
        const {
            required: aRequired,
            properties: aProperties,
            type: aType,
            ...aRest
        } = a;
        const {
            required: bRequired,
            properties: bProperties,
            type: bType,
            ...bRest
        } = b;
        if (
            Object.keys(aRest).length === 0 &&
            Object.keys(bRest).length === 0 &&
            aType === 'object' &&
            bType === 'object'
        ) {
            return {
                required: [...(aRequired ?? []), ...(bRequired ?? [])],
                type: 'object',
                properties: { ...aProperties, ...bProperties },
            };
        }
    }
    return {
        allOf: [a, b],
    };
}

export function mergeAllOfs(schemas: JSONSchema[]): JSONSchema {
    if (schemas.length === 1) {
        return schemas[0];
    }
    const [a, b, ...rest] = schemas;
    return mergeAllOfs([mergeAllOf(a, b), ...rest]);
}
