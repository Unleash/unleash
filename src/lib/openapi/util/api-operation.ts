import { OpenAPIV3 } from 'openapi-types';
import { OpenApiTag } from './openapi-tags';

type DeprecatedOpenAPITag =
    // Deprecated tag names. Please use a tag from the OpenAPITag type instead.
    //
    // These tag names were the original ones we used for OpenAPI, but they
    // turned out to be too generic and/or didn't match the new tag naming
    // schema. Because we require our operations to have one of a predefined set
    // of values, it would be breaking change to remove them completely.
    'client' | 'other' | 'auth' | 'admin';

export interface ApiOperation<Tag = OpenApiTag | DeprecatedOpenAPITag>
    extends Omit<OpenAPIV3.OperationObject, 'tags'> {
    operationId: string;
    tags: [Tag];
}
