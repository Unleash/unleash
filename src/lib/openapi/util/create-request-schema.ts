import { OpenAPIV3 } from 'openapi-types';

export const createRequestSchema = (
    schemaName: string,
): OpenAPIV3.RequestBodyObject => {
    return {
        description: schemaName,
        required: true,
        content: {
            'application/json': {
                schema: {
                    $ref: `#/components/schemas/${schemaName}`,
                },
            },
        },
    };
};
