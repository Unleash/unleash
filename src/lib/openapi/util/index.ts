export * from './api-operation';
export * from './create-request-schema';
export * from './create-response-schema';
export * from './openapi-tags';
export * from './standard-responses';
