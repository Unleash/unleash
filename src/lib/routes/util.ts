import joi from 'joi';
import { Response } from 'express';
import { Logger } from '../logger';
import BaseError from '../error/base-error';

export const customJoi = joi.extend((j) => ({
    type: 'isUrlFriendly',
    base: j.string(),
    messages: {
        'isUrlFriendly.base': '{{#label}} must be URL friendly',
    },
    validate(value, helpers) {
        // Base validation regardless of the rules applied
        if (encodeURIComponent(value) !== value) {
            // Generate an error, state and options need to be passed
            return { value, errors: helpers.error('isUrlFriendly.base') };
        }
        return undefined;
    },
}));

export const nameType = customJoi.isUrlFriendly().min(1).max(100).required();

export const handleErrors: (
    res: Response,
    logger: Logger,
    error: Error,
) => void = (res, logger, error) => {
    logger.warn(error.message);
    // @ts-expect-error
    // eslint-disable-next-line no-param-reassign
    error.isJoi = true;

    if (error instanceof BaseError) {
        return res.status(error.statusCode).json(error).end();
    }

    switch (error.name) {
        case 'ValidationError':
            return res.status(400).json(error).end();
        case 'BadDataError':
            return res.status(400).json(error).end();
        case 'OwaspValidationError':
            return res.status(400).json(error).end();
        case 'PasswordUndefinedError':
            return res.status(400).json(error).end();
        case 'MinimumOneEnvironmentError':
            return res.status(400).json(error).end();
        case 'InvalidTokenError':
            return res.status(401).json(error).end();
        case 'NoAccessError':
            return res.status(403).json(error).end();
        case 'UsedTokenError':
            return res.status(403).json(error).end();
        case 'InvalidOperationError':
            return res.status(403).json(error).end();
        case 'IncompatibleProjectError':
            return res.status(403).json(error).end();
        case 'OperationDeniedError':
            return res.status(403).json(error).end();
        case 'NotFoundError':
            return res.status(404).json(error).end();
        case 'NameExistsError':
            return res.status(409).json(error).end();
        case 'FeatureHasTagError':
            return res.status(409).json(error).end();
        case 'RoleInUseError':
            return res.status(400).json(error).end();
        case 'ProjectWithoutOwnerError':
            return res.status(409).json(error).end();
        default:
            logger.error('Server failed executing request', error);
            return res.status(500).end();
    }
};
