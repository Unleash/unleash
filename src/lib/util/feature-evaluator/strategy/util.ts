import * as murmurHash3 from 'murmurhash3js';

export default function normalizedValue(id: string, groupId: string, normalizer = 100): number {
  return (murmurHash3.x86.hash32(`${groupId}:${id}`) % normalizer) + 1;
}
