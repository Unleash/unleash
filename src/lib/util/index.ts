export * from './timer';
export * from './semver';
export * from './version';
export * from './is-email';
export * from './segments';
export * from './anonymise';
export * from './constants';
export * from './isDefined';
export * from './omit-keys';
export * from './random-id';
export * from './snakeCase';
export * from './map-values';
export * from './time-utils';
export * from './collect-ids';
export * from './parseEnvVar';
export * from './rewriteHTML';
export * from './extract-user';
export * from './flag-resolver';
export * from './metrics-helper';
export * from './validateOrigin';
export * from './anyEventEmitter';
export * from './format-base-uri';
export * from './load-index-html';
export * from './findPublicFolder';
export * from './generateImageUrl';
export * from './ensureStringValue';
export * from './graceful-shutdown';
export * from './validateJsonString';
export * from './arraysHaveSameItems';
export * from './constantTimeCompare';
export * from './collapseHourlyMetrics';
export * from './offline-unleash-client';
