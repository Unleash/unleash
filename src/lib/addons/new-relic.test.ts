import {
    FEATURE_ARCHIVED,
    FEATURE_CREATED,
    FEATURE_ENVIRONMENT_DISABLED,
    type IFlagResolver,
    type IAddonConfig,
    type IEvent,
    serializeDates,
    type IFlagKey,
} from '../types';
import type { Logger } from '../logger';

import NewRelicAddon, { type INewRelicParameters } from './new-relic';

import noLogger from '../../test/fixtures/no-logger';
import { gunzip } from 'node:zlib';
import { promisify } from 'util';
import type { IntegrationEventsService } from '../services';

const asyncGunzip = promisify(gunzip);

let fetchRetryCalls: any[] = [];
const registerEventMock = jest.fn();

const INTEGRATION_ID = 1337;
const ARGS: IAddonConfig = {
    getLogger: noLogger,
    unleashUrl: 'http://some-url.com',
    integrationEventsService: {} as IntegrationEventsService,
    flagResolver: { isEnabled: (expName: IFlagKey) => false } as IFlagResolver,
    eventBus: {} as any,
};

jest.mock(
    './addon',
    () =>
        class Addon {
            logger: Logger;

            constructor(definition, { getLogger }) {
                this.logger = getLogger('addon/test');
                fetchRetryCalls = [];
            }

            async fetchRetry(url, options, retries, backoff) {
                fetchRetryCalls.push({
                    url,
                    options,
                    retries,
                    backoff,
                });
                return Promise.resolve({ ok: true, status: 200 });
            }

            async registerEvent(event) {
                return registerEventMock(event);
            }
        },
);

const defaultParameters = {
    url: 'fakeUrl',
    licenseKey: 'fakeLicenseKey',
} as INewRelicParameters;

const defaultEvent = {
    id: 1,
    createdAt: new Date(),
    type: FEATURE_CREATED,
    createdBy: 'some@user.com',
    createdByUserId: -1337,
    featureName: 'some-toggle',
    data: {
        name: 'some-toggle',
        enabled: false,
        strategies: [{ name: 'default' }],
    },
} as IEvent;

const makeAddHandleEvent = (event: IEvent, parameters: INewRelicParameters) => {
    const addon = new NewRelicAddon(ARGS);

    return () => addon.handleEvent(event, parameters, INTEGRATION_ID);
};

describe('New Relic integration', () => {
    beforeEach(() => {
        registerEventMock.mockClear();
    });

    test.each([
        {
            partialEvent: { type: FEATURE_CREATED },
            test: '$type toggle',
        },
        {
            partialEvent: {
                type: FEATURE_ARCHIVED,
                data: {
                    name: 'some-toggle',
                },
            },
            test: 'FEATURE_ARCHIVED toggle with project info',
        },
        {
            partialEvent: {
                type: FEATURE_ARCHIVED,
                project: 'some-project',
                data: {
                    name: 'some-toggle',
                },
            },
            test: 'FEATURE_ARCHIVED with project info',
        },
        {
            partialEvent: {
                type: FEATURE_ENVIRONMENT_DISABLED,
                environment: 'development',
            },
            test: 'toggled environment',
        },
        {
            partialEvent: {
                type: FEATURE_ENVIRONMENT_DISABLED,
                environment: 'development',
            },
            partialParameters: {
                customHeaders: `{ "MY_CUSTOM_HEADER": "MY_CUSTOM_VALUE" }`,
            },
            test: 'customHeaders in headers when calling service',
        },
        {
            partialEvent: {
                type: FEATURE_ENVIRONMENT_DISABLED,
                environment: 'development',
            },
            partialParameters: {
                bodyTemplate:
                    '{\n  "eventType": "{{event.type}}",\n  "createdBy": "{{event.createdBy}}"\n}',
            },
            test: 'custom body template',
        },
    ] as Array<{
        partialEvent: Partial<IEvent>;
        partialParameters?: Partial<INewRelicParameters>;
        test: String;
    }>)(
        'Should call New Relic Event API for $test',
        async ({ partialEvent, partialParameters }) => {
            const event = {
                ...defaultEvent,
                ...partialEvent,
            };

            const parameters = {
                ...defaultParameters,
                ...partialParameters,
            };

            const handleEvent = makeAddHandleEvent(event, parameters);

            await handleEvent();
            expect(fetchRetryCalls.length).toBe(1);

            const { url, options } = fetchRetryCalls[0];
            const jsonBody = JSON.parse(
                (await asyncGunzip(options.body)).toString(),
            );

            expect(url).toBe(parameters.url);
            expect(options.method).toBe('POST');
            expect(options.headers['Api-Key']).toBe(parameters.licenseKey);
            expect(options.headers['Content-Type']).toBe('application/json');
            expect(options.headers['Content-Encoding']).toBe('gzip');
            expect(options.headers).toMatchSnapshot();

            expect(jsonBody.eventType).toBe('UnleashServiceEvent');
            expect(jsonBody.unleashEventType).toBe(event.type);
            expect(jsonBody.featureName).toBe(event.data.name);
            expect(jsonBody.environment).toBe(event.environment);
            expect(jsonBody.createdBy).toBe(event.createdBy);
            expect(jsonBody.createdByUserId).toBe(event.createdByUserId);
            expect(jsonBody.createdAt).toBe(event.createdAt.getTime());
        },
    );

    test('Should call registerEvent', async () => {
        const handleEvent = makeAddHandleEvent(defaultEvent, defaultParameters);

        await handleEvent();

        expect(registerEventMock).toHaveBeenCalledTimes(1);
        expect(registerEventMock).toHaveBeenCalledWith({
            integrationId: INTEGRATION_ID,
            state: 'success',
            stateDetails:
                'New Relic Events API request was successful with status code: 200.',
            event: serializeDates(defaultEvent),
            details: {
                url: defaultParameters.url,
                body: {
                    eventType: 'UnleashServiceEvent',
                    unleashEventType: defaultEvent.type,
                    featureName: defaultEvent.featureName,
                    environment: defaultEvent.environment,
                    createdBy: defaultEvent.createdBy,
                    createdByUserId: defaultEvent.createdByUserId,
                    createdAt: defaultEvent.createdAt.getTime(),
                    ...defaultEvent.data,
                },
            },
        });
    });
});
