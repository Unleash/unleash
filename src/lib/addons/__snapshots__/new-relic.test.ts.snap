// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Relic integration Should call New Relic Event API for $type toggle 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
}
`;

exports[`New Relic integration Should call New Relic Event API for FEATURE_ARCHIVED toggle with project info 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
}
`;

exports[`New Relic integration Should call New Relic Event API for FEATURE_ARCHIVED with project info 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
}
`;

exports[`New Relic integration Should call New Relic Event API for custom body template 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
}
`;

exports[`New Relic integration Should call New Relic Event API for customHeaders in headers when calling service 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
  "MY_CUSTOM_HEADER": "MY_CUSTOM_VALUE",
}
`;

exports[`New Relic integration Should call New Relic Event API for toggled environment 1`] = `
{
  "Api-Key": "fakeLicenseKey",
  "Content-Encoding": "gzip",
  "Content-Type": "application/json",
}
`;
